File: 2023/day2/solution.py
from lib.parse import parse_strings

class Game:

    def __init__(self, input: str) -> None:
        self.RED = "red"
        self.BLUE = "blue"
        self.GREEN = "green"

        raw_label, raw_sets = input.split(": ")
        self.id = int(raw_label.split(" ")[1])
        
        self.rounds = []
        sets = raw_sets.split("; ")
        for s in sets:
            mapping = {self.RED: 0, self.BLUE: 0, self.GREEN: 0}
            colors = s.split(", ")
            for c in colors:
                mapping[c.split(" ")[1]] = int(c.split(" ")[0])
            self.rounds.append(mapping)

    def possible(self, red_lim: int, green_lim: int, blue_lim: int) -> bool:
        for r in self.rounds:
            if r[self.RED] > red_lim or r[self.BLUE] > blue_lim or r[self.GREEN] > green_lim:
                return False
        return True
    
    def min_needed(self) -> int:
        red_max, green_max, blue_max = 0, 0, 0
        for r in self.rounds:
            red_max = max(r[self.RED], red_max)
            green_max = max(r[self.GREEN], green_max)
            blue_max = max(r[self.BLUE], blue_max)
        return red_max * green_max * blue_max


def get_valid_game_sum() -> int:
    data = parse_strings("2023/day2/input.txt")
    total = 0
    for d in data:
        g = Game(input=d)
        if g.possible(red_lim=12, green_lim=13, blue_lim=14):
            total += g.id

    return total

def get_min_game_sum() -> int:
    data = parse_strings("2023/day2/input.txt")
    total = 0
    for d in data:
        g = Game(input=d)
        total += g.min_needed()
    return total

print(get_valid_game_sum())
print(get_min_game_sum())

File: 2023/day2/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Game struct {
	ID     int
	Rounds []map[string]int
}

func getInput(filePath string) ([]Game, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return []Game{}, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	res := []Game{}

	for scanner.Scan() {
		line := scanner.Text()
		data := strings.Split(line, ": ")

		id, _ := strconv.Atoi(strings.Split(data[0], " ")[1])
		rounds := strings.Split(data[1], "; ")
		g := Game{ID: id}

		for _, r := range rounds {
			cubes := strings.Split(r, ", ")
			nextRound := map[string]int{}
			for _, c := range cubes {
				colorData := strings.Split(c, " ")
				amt, _ := strconv.Atoi(colorData[0])
				color := colorData[1]
				nextRound[color] = amt
			}
			g.Rounds = append(g.Rounds, nextRound)
		}
		res = append(res, g)
	}
	return res, nil
}

func getValidRounds(games []Game) int {
	sum := 0
	for _, g := range games {
		possible := true
		for _, r := range g.Rounds {
			STANDARD := map[string]int{
				"red":   12,
				"green": 13,
				"blue":  14,
			}
			for color, amt := range r {
				STANDARD[color] -= amt
				if STANDARD[color] < 0 {
					possible = false
				}
			}
		}
		if possible {
			sum += g.ID
		}
	}
	return sum
}

func getMinimumCounts(games []Game) int {
	powers := 0
	for _, g := range games {
		MINIMUM := map[string]int{
			"red":   0,
			"green": 0,
			"blue":  0,
		}
		for _, r := range g.Rounds {
			for color, amt := range r {
				MINIMUM[color] = max(MINIMUM[color], amt)
			}
		}
		powers += MINIMUM["red"] * MINIMUM["green"] * MINIMUM["blue"]
	}
	return powers
}

func main() {
	games, err := getInput("input.txt")
	if err != nil {
		panic(err)
	}
	part1 := getValidRounds(games)
	fmt.Printf("Part 1: %d\n", part1)

	part2 := getMinimumCounts(games)
	fmt.Printf("Part 2: %d\n", part2)
}
